R/leave_location_out_cv.R

```diff
@@ -0,0 +1,193 @@
+#'
+#' @export
+leave_location_out_cv <- function(data, location, v = length(unique(location)),
+                                  pool = 0, ...) {
+  location <- tidyselect::eval_select(rlang::enquo(location), data = data)
+
+  if (is_empty(location)) {
+    rlang::abort("`location` is required and must be variables in `data`.")
+  }
+  location <- make_location(getElement(data, location), pool = pool)
+  n_locations <- length(unique(location))
+
+  if (missing(v)) {
+    v <- n_locations
+  } else if (!is.numeric(v) || length(v) != 1) {
+    rlang::abort("`v` must be a single integer.")
+  } else if (v > n_locations) {
+    rlang::warn(paste0(
+      "Fewer than ", v, " locations available for sampling; setting v to ",
+      n_locations, "."
+    ))
+    v <- n_locations
+  }
+
+  split_objs <- llo_splits(
+    data = data, location = location, v = v, pool = pool
+  )
+
+  ## We remove the holdout indices since it will save space and we can
+  ## derive them later when they are needed.
+
+  split_objs$splits <- map(split_objs$splits, rm_out)
+
+  ## Save some overall information
+
+  cv_att <- list(v = v)
+
+  new_rset(
+    splits = split_objs$splits,
+    ids = split_objs[, grepl("^id", names(split_objs))],
+    attrib = cv_att,
+    subclass = c("leave_location_out_cv", "rset")
+  )
+}
+
+llo_splits <- function(data, location, v, pool = 0.1) {
+
+  n <- nrow(data)
+  locations <- tibble::tibble(
+    idx = seq_len(n),
+    location = location
+  )
+
+  folds <- tibble::tibble(
+    location = unique(locations$location),
+    fold = sample(rep(seq_len(v), length.out = length(unique(locations$location))))
+  )
+
+  locations <- merge(locations, folds, by = c("location"))
+  indices <- split_unnamed(locations$idx, locations$fold)
+  indices <- lapply(indices, default_complement, n = n)
+  split_objs <- purrr::map(
+    indices,
+    make_splits,
+    data = data,
+    class = "llo_split"
+  )
+  tibble::tibble(
+    splits = split_objs,
+    id = names0(length(split_objs), "Fold")
+  )
+}
+
+#' @export
+print.leave_location_out_cv <- function(x, ...) {
+    cat("# ", pretty(x), "\n")
+    class(x) <- class(x)[!(class(x) %in% c("leave_location_out_cv", "rset"))]
+    print(x, ...)
+}
+
+#' Create or Modify Location Variables
+#'
+#' @details
+#' For categorical inputs, the function will find levels of `x` than
+#'   occur in the data with percentage less than `pool`. The values from
+#'   these groups will be pooled into a single category (if they make up
+#'   a greater than `pool` percentage of the data) or combined with the smallest
+#'   extant category (if not).
+#'
+#' @param x An input vector.
+#' @param pool A proportion of data used to determine if a particular group is
+#'   too small and should be pooled into another group.
+#' @return  A factor vector.
+make_location <- function(x, pool = .1) {
+
+  n <- length(x)
+  x <- factor(x)
+  xtab <- sort(table(x))
+  pcts <- xtab / n
+
+  if (all(pcts < pool) || sum(pcts > pool) == 1) {
```

Comment:
`pool` defaults to 0 in `leave_location_out_cv()`, because it doesn't seem unreasonable to me that someone might have 10+ locations and want to iterate through each of them (say you're using each island in the Galapagos as a location). There's probably _a_ minimum size that makes sense as a default, but I'm not sure what it would be.

The intent of the second clause is to avoid having a single small location come out greater than `pool`, and then a massive pooled group (since if pool = 0.1, you could have a 90% pool/10% user-defined location split). Right now, the way this is written, if one location makes up 51% of your data you can never use this function. That's probably not ideal, but I'm not sure.